pkg/operator/ceph/cluster/cluster.go

```diff
@@ -290,48 +289,33 @@ func clusterChanged(oldCluster, newCluster cephv1.ClusterSpec, clusterRef *clust
 	return false, ""
 }
 
-func (c *cluster) unsetOrchestrationRunning() {
-	c.orchRunMux.Lock()
+// unsetOrchestrationStatus resets the orchestration flags to its original status when
+// the orchestration is done
+func (c *cluster) unsetOrchestrationStatus() {
+	c.orchMux.Lock()
 	c.orchestrationRunning = false
-	c.orchRunMux.Unlock()
-}
-
-func (c *cluster) setOrchestrationPending() {
-	c.orchPenMux.Lock()
 	c.orchestrationPending = true
-	c.orchPenMux.Unlock()
-}
-
-func (c *cluster) getOrchestrationRunning() bool {
-	c.orchRunMux.Lock()
-	defer c.orchRunMux.Unlock()
-	return c.orchestrationRunning
+	c.orchMux.Unlock()
 }
 
-func (c *cluster) getOrchestrationPending() bool {
-	c.orchPenMux.Lock()
-	defer c.orchPenMux.Unlock()
-	return c.orchestrationPending
-}
-
-// make checking and setting orchestrationRunning "atomic"
-func (c *cluster) checkSetOrchestrationRunning() bool {
-	defer c.orchRunMux.Unlock()
-	c.orchRunMux.Lock()
-	if c.orchestrationRunning == false {
+// checkSetOrchestrationStatus is responsible to do orchestration as long as there is a request pending
+// the thread local variable init indicates if the calling function is the initial one which started the
+// orchestration.
+func (c *cluster) checkSetOrchestrationStatus(init bool) bool {
+	c.orchMux.Lock()
+	defer c.orchMux.Unlock()
+	// check if there is an orchestration pending currently
+	// NOTE: c.orchestrationPending is initialized as true to get the very first request running
+	if c.orchestrationPending == true && c.orchestrationRunning == false {
+		c.orchestrationPending = false
 		c.orchestrationRunning = true
-		return false
-	}
-	return true
-}
-
-// make checking and unsetting orchestrationPending "atomic"
-func (c *cluster) checkUnsetOrchestrationPending() bool {
-	defer c.orchPenMux.Unlock()
-	c.orchPenMux.Lock()
-	if c.orchestrationPending == true {
+		// allow to run an orchestration
+		return true
+	} else if c.orchestrationPending == true && c.orchestrationRunning == true && init == true {
```

Comment:
This condition seems dangerous (and unused?). If an orchestration is pending, and another thread is running orchestration, and this is not the first loop, go ahead and run an orchestration anyway?

Reply:
This condition is used by the running orchestration thread (-> `init` is set) to indicate that this is the initial `createInstance` call.



| Thread A | Thread B | env | comment |
| ------------ |------------ | -------------- | ------------ |
| createInstance| | pend = T, run = F | |
| for true (pend = T & run = F, init = false) | | | |
| set init = true | | pend = F, run = T | init is thread-local|
| _work_ | createInstance | pend = F, run = T | |
| _work_ | for false (pend = F & run = T, init = false) | pend = T, run = T | Thread B set pend|
| for true (pend = T & run = T, init = true) | | pend = F, run = T | Thread A starts **another** loop | as it is the initial `createInstance`-Instance and as there is a pending orchestration |